File: goldenflow/domains/finance.py
from __future__ import annotations
import re
import polars as pl
from goldenflow.config.schema import GoldenFlowConfig, TransformSpec
from goldenflow.domains.base import DomainPack
from goldenflow.transforms import register_transform

@register_transform(name="account_mask", input_types=["string"], auto_apply=False, priority=50, mode="series")
def account_mask(series: pl.Series) -> pl.Series:
    """Mask account numbers showing only last 4 digits."""
    def _mask(val):
        if val is None: return None
        digits = re.sub(r"\D", "", str(val))
        if len(digits) < 4: return val
        return "*" * (len(digits) - 4) + digits[-4:]
    return series.map_elements(_mask, return_dtype=pl.Utf8)

@register_transform(name="cusip_format", input_types=["string"], auto_apply=False, priority=50, mode="series")
def cusip_format(series: pl.Series) -> pl.Series:
    """Standardize CUSIP identifiers (9 chars, uppercase)."""
    def _fmt(val):
        if val is None: return None
        return val.strip().upper()[:9]
    return series.map_elements(_fmt, return_dtype=pl.Utf8)

PACK = DomainPack(
    name="finance",
    description="Account masking, currency standardization, CUSIP/ISIN formatting",
    transforms=["account_mask", "cusip_format", "currency_strip", "date_iso8601"],
    default_config=GoldenFlowConfig(
        transforms=[
            TransformSpec(column="account_number", ops=["account_mask"]),
            TransformSpec(column="amount", ops=["currency_strip"]),
            TransformSpec(column="transaction_date", ops=["date_iso8601"]),
        ]
    ),
)

File: packages/goldenflow-js/src/core/domains/finance.ts
import type { ColumnValue, DomainPack } from "../types.js";
import { makeConfig } from "../types.js";
import { registerTransform } from "../transforms/registry.js";

function accountMask(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const digits = v.replace(/\D/g, "");
    if (digits.length < 4) return v;
    return "*".repeat(digits.length - 4) + digits.slice(-4);
  });
}

function cusipFormat(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    return v.trim().toUpperCase().slice(0, 9);
  });
}

registerTransform(
  { name: "account_mask", inputTypes: ["string"], priority: 50, mode: "series" },
  accountMask,
);
registerTransform(
  { name: "cusip_format", inputTypes: ["string"], priority: 50, mode: "series" },
  cusipFormat,
);

export const PACK: DomainPack = {
  name: "finance",
  description: "Account masking, currency standardization, CUSIP/ISIN formatting",
  transforms: ["account_mask", "cusip_format", "currency_strip", "date_iso8601"],
  defaultConfig: makeConfig({
    transforms: [
      { column: "account_number", ops: ["account_mask"] },
      { column: "amount", ops: ["currency_strip"] },
      { column: "transaction_date", ops: ["date_iso8601"] },
    ],
  }),
};
